text/0031-configuration.md

```diff
@@ -0,0 +1,280 @@
+    Ok(toml) => match fs::File::create("config.toml") {
+        Ok(mut file) => {
+            if let Err(e) = file.write_all(toml.as_bytes()) {
+                panic!("Error writing .toml config file: {:?}", e);
+            }
+        }
+        Err(e) => {
+            panic!("Error creating .toml config file: {:?}", e);
+        }
+    },
+    Err(e) => {
+        panic!("Error serializing .toml config file: {:?}", e);
+    }
+}
+```
+
+### Sub-configuration
+
+It is also very easy to create sub-configurations by nesting configuration builders and configurations.
+
+`config.toml`
+```toml
+[snapshot]
+[snapshot.local]
+meta_file_path  = "./data/snapshot/local/mainnet.snapshot.meta"
+state_file_path = "./data/snapshot/local/mainnet.snapshot.state"
+[snapshot.global]
+file_path       = "./data/snapshot/global/mainnet.txt"
+```
+
+`config.rs`
+```rust
+#[derive(Default, Deserialize)]
+pub struct LocalSnapshotConfigBuilder {
+    meta_file_path: Option<String>,
+    state_file_path: Option<String>,
+}
+
+#[derive(Default, Deserialize)]
+pub struct GlobalSnapshotConfigBuilder {
+    file_path: Option<String>,
+}
+
+#[derive(Default, Deserialize)]
+pub struct SnapshotConfigBuilder {
+    local: LocalSnapshotConfigBuilder,
+    global: GlobalSnapshotConfigBuilder,
+}
+
+impl SnapshotConfigBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    // Setters
+
+    pub fn build(self) -> SnapshotConfig {
+        SnapshotConfig {
+            local: LocalSnapshotConfig {
+                meta_file_path: self
+                    .local
+                    .meta_file_path
+                    .unwrap_or(DEFAULT_LOCAL_SNAPSHOT_META_FILE_PATH.to_string()),
+                state_file_path: self
+                    .local
+                    .state_file_path
+                    .unwrap_or(DEFAULT_LOCAL_SNAPSHOT_STATE_FILE_PATH.to_string()),
+            },
+            global: GlobalSnapshotConfig {
+                file_path: self
+                    .global
+                    .file_path
+                    .unwrap_or(DEFAULT_GLOBAL_SNAPSHOT_FILE_PATH.to_string()),
+            },
+        }
+    }
+}
+#[derive(Clone)]
+pub struct LocalSnapshotConfig {
+    meta_file_path: String,
+    state_file_path: String,
+}
+
+#[derive(Clone)]
+pub struct GlobalSnapshotConfig {
+    file_path: String,
+}
+
+#[derive(Clone)]
+pub struct SnapshotConfig {
+    local: LocalSnapshotConfig,
+    global: GlobalSnapshotConfig,
+}
+
+// Impl
+
+```
+
+# Drawbacks
+
+No specific drawback with this approach.
+
+# Rationale and alternatives
+
+Many configuration formats are usually considered:
+- [TOML](https://github.com/toml-lang/toml)
+- [JSON](https://www.json.org/json-en.html)
+- [HOCON](https://github.com/lightbend/config/blob/master/HOCON.md)
+- [YAML](https://yaml.org/)
+- [XML](https://www.w3.org/XML/)
+
+This RFC and its expected implementation are choosing `TOML` as configuration format because it is the preferred option
```

Comment:
Be aware that `TOML` does somewhat restrict the data that can be serialised and cannot be used to serialise arbitrarily nested types. I'd recommend specifying that `RON` (https://github.com/ron-rs/ron) or `JSON` should be the preferred backup should `TOML` not be sufficient.

Reply:
What is the limitation on nested types ? I'm actually having a section about that and it works fine ?

Reply:
I can mention `RON`. I think we're just giving a default to the community but that we'll end up adapting if need be. I'll add a note!

Reply:
TOML only supports about 2 layers of nesting due to its non-recursive syntax.

Reply:
Ahhh didn't know that! Let's see where it leads us. I'll let a comment. Switching format with serde will be very easy anyway.